compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/FuseAndHoistParallelLoops.cpp

```diff
@@ -113,15 +114,63 @@ struct FuseTileableDestinationProducers final
   }
 };
 
+struct FuseTilableForallConsumers final
+    : OpInterfaceRewritePattern<TilingInterface> {
+  using OpInterfaceRewritePattern::OpInterfaceRewritePattern;
+  LogicalResult matchAndRewrite(TilingInterface tilableOp,
+                                PatternRewriter &rewriter) const override {
+    // Currently consumer fusion requires DPS, and we don't want to fuse through
+    // inits anyway.
+    auto dpsOp = dyn_cast<DestinationStyleOpInterface>(*tilableOp);
+    if (!dpsOp) {
+      return failure();
+    }
+
+    tensor::ParallelInsertSliceOp producerSlice;
+    for (auto operand : dpsOp.getDpsInputs()) {
+      auto forallProducer = operand.getDefiningOp<scf::ForallOp>();
+      if (!forallProducer) {
+        continue;
+      }
+      Value iterArg = forallProducer.getTiedBlockArgument(
+          forallProducer.getTiedOpOperand(cast<OpResult>(operand)));
+
+      for (auto user : iterArg.getUsers()) {
+        auto sliceOp = dyn_cast<tensor::ParallelInsertSliceOp>(user);
+        if (sliceOp && sliceOp.getDest() == iterArg) {
+          producerSlice = sliceOp;
+          break;
+        }
+      }
+      if (producerSlice) {
+        break;
+      }
+    }
+
+    if (!producerSlice) {
+      return failure();
+    }
+
+    FailureOr<scf::SCFFuseConsumerOfSliceResult> fuseConsumerResults =
+        scf::tileAndFuseConsumerOfSlice(rewriter, producerSlice);
+    if (failed(fuseConsumerResults)) {
+      return failure();
+    }
+    return success();
+  }
+};
+
 void FuseAndHoistParallelLoopsPass::runOnOperation() {
   MLIRContext *context = &getContext();
   RewritePatternSet patterns(context);
 
   // These two patterns are run to a fixed point, allowing fusion within
```

Comment:
Maybe drop the fixed point here. Do you need a fixed point? I can see how fuse for alls is a fixed point, but I think you want to control the order in which the forall fusion and fusion of consumers/producers kicks in. Those two probably needs to be explicit worklist.

Reply:
I moved it to a separate pattern application. I'm planning to move it to an explicit worklist in a follow up if that's ok.